enterprise/server/remoteauth/remoteauth.go

```diff
@@ -0,0 +1,182 @@
+package remoteauth
+
+import (
+	"context"
+	"flag"
+	"net/http"
+
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/util/authutil"
+	"github.com/buildbuddy-io/buildbuddy/server/util/claims"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/hash"
+	"github.com/buildbuddy-io/buildbuddy/server/util/lru"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+
+	authpb "github.com/buildbuddy-io/buildbuddy/proto/auth"
+)
+
+const (
+	jwtCacheSize = 10_000
+)
+
+var (
+	authHeaders = []string{authutil.APIKeyHeader}
+
+	remoteAuthTarget = flag.String("auth.remote_auth_target", "grpcs://remote.buildbuddy.dev", "The gRPC target of the remote authentication API.")
+)
+
+func NewRemoteAuthenticator() (*RemoteAuthenticator, error) {
+	conn, err := grpc_client.DialSimpleWithoutPooling(*remoteAuthTarget)
+	if err != nil {
+		return nil, err
+	}
+	return newRemoteAuthenticator(conn)
+}
+
+func newRemoteAuthenticator(conn *grpc.ClientConn) (*RemoteAuthenticator, error) {
+	config := &lru.Config[string]{
+		MaxSize: jwtCacheSize,
+		SizeFn:  func(v string) int64 { return 1 },
+	}
+	cache, err := lru.NewLRU(config)
+	if err != nil {
+		return nil, err
+	}
+	return &RemoteAuthenticator{
+		authClient: authpb.NewAuthServiceClient(conn),
+		cache:      cache,
+	}, nil
+}
+
+type RemoteAuthenticator struct {
+	authClient authpb.AuthServiceClient
+	cache      interfaces.LRU[string]
+}
+
+// Admin stuff unsupported in remote authenticator.
+func (a *RemoteAuthenticator) AdminGroupID() string {
+	return ""
+}
+
+// TODO(iain): control via flag if needed.
+func (a *RemoteAuthenticator) AnonymousUsageEnabled(ctx context.Context) bool {
+	return false
+}
+
+// Unsupported in remote authenticator.
+func (a *RemoteAuthenticator) PublicIssuers() []string {
+	return []string{}
+}
+
+func (a *RemoteAuthenticator) Login(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP login unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) Logout(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP logout unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) Auth(w http.ResponseWriter, r *http.Request) error {
+	return status.UnimplementedError("HTTP auth unsupported with remote authentication")
+}
+
+func (a *RemoteAuthenticator) AuthenticatedHTTPContext(w http.ResponseWriter, r *http.Request) context.Context {
+	return r.Context()
+}
+
+func (a *RemoteAuthenticator) SSOEnabled() bool {
+	return false
+}
+
+func (a *RemoteAuthenticator) AuthenticatedGRPCContext(ctx context.Context) context.Context {
+	key, err := hashAuthHeaders(ctx)
+	if err != nil {
+		jwt, err := a.authenticate(ctx)
```

Comment:
Why do we call authenticate here if hashAuthHeaders fails?

Reply:
I just figured that was a safe backup, but switched it to just using ctx in that case.